// identification for single matching ad
message AdId {
  required int32 campaign_id          = 1;        // The campaign's ID.
  required int32 placement_id         = 2;        // The placement's ID
  required int32 creative_id          = 3;        // The creative's ID.
}

/* ADS-6296 Additional attributes in RTB */
message Geo {
  /* openRTB-compliant fields */
  optional float lat         = 1;       // Latitude
  optional float lon         = 2;       // Longitude
  optional string country    = 3;       // The user's Country
  optional string city       = 4;       // The user's city
  optional string zip        = 5;       // The user's zip/postal code
  optional int32 type        = 6;       // Source of geo data
  /* openx-specific fields */
  optional string continent  = 7;       // The user's continent
  optional string state      = 8;       // The user's state
  optional int32  dma        = 9;       // The user's dma
}

message Device {
  /* openRTB-compliant fields */
  optional string didsha1          = 1;   // SHA1 hashed device identifier 
  optional int32 dnt               = 2;   // Do not Track indicator. If "0", 
                                          // then do not track Is set to 
                                          // false, if "1", then do no track 
                                          // is set to true in browser.
  optional string ua               = 3;   // Browser user agent string
  optional string ip               = 4;   // IPv4 address closest to device
  optional string carrier          = 5;   // Mobile Country Code-Mobile 
                                          // Network Code
  optional string os               = 6;   // Device operating system
  optional string osv              = 7;   // Device operating system version
  repeated string language         = 8;   // Browser language
  optional string make             = 9;   // Device make
  optional string model            = 10;  // Device model
  optional string connectiontype   = 11;  // Detected data connection type 
                                          // for the device Values are wifi, 
                                          // cell, cell-2G, cell-3G, cell-4G
  optional string deprecated_api   = 12;  // Deprecated. api in the format of
                                          // repeated int32 should be used.
  optional Geo geo                 = 13;  // Geo object
  /* openx-specific fields */
  optional string browser          = 14;  // Browser name
  optional string browser_version  = 15;  // Browser version
  repeated int32  api              = 16;  // List of supported API frameworks
}

message BidRequest {

  // ------------------------------------------------------------------------
  // ***    REQUIRED FIELDS                                               ***
  // ------------------------------------------------------------------------
  // These fields are guaranteed to be set by the OpenX RTB system and passed
  // to the bidder in the bid request.

  // OPENX identifiers
  required int32 api_version         = 1;        // This value represents the
                                                 // version of the API the
                                                 // Market is using to
                                                 // communicate with the
                                                 // real-time bidder.
  required string auction_id         = 2;        // This is a unique value
                                                 // generated by the Market
                                                 // and used to identify the
                                                 // auction request.

  // ADVERTISER identifiers
  repeated AdId   matching_ad_ids    = 3;        // list of ads which match
                                                 // for this request

  // PUBLISHER identifiers
  required string pub_website_id     = 4;        // The ID of the publisher's
                                                 // website

  // AD size
  required int32  ad_height          = 5;        // The height of the ad
                                                 // position to fill
  required int32  ad_width           = 6;        // The width of the ad
                                                 // position to fill

  // ------------------------------------------------------------------------
  // ***    OPTIONAL FIELDS                                               ***
  // ------------------------------------------------------------------------
  // These fields are optional and may be set by the OpenX RTB system and
  // passed to the bidder in the bid request.

  // USER related fields
  optional string user_cookie_id     = 7;        // A unique ID representing the
                                                 // user.  To protect the user's
                                                 // identity this value is
                                                 // different for each bidder.
  optional bytes  user_ip_address    = 8;        // The IP address of the user
  optional int32  user_screen_height = 9;        // The height of the
                                                 // user's screen
  optional int32  user_screen_width  = 10;       // The width of the
                                                 // user's screen
  optional string user_geo_country   = 11;       // The user's country
  optional string user_geo_state     = 12;       // The user's state
  optional int32  user_geo_dma       = 13;       // The user's dma
  optional string user_agent         = 14;       // The user's browser, version,
                                                 // and operating system.
  repeated string user_lang          = 15;       // The language preferences set
                                                 // in the user's browser

  // PUBLISHER related fields
  optional string url                = 16;       // The URL the ad request is
                                                 // coming from.
  optional string http_referer       = 17;       // The HTTP_REFERER for the
                                                 // page generating the ad
                                                 // request
  optional int32  ox_cat_tier_1      = 18;       // The top level category of
                                                 // the page's content.
  optional int32  ox_cat_tier_2      = 19;       // The second level category of
                                                 // the page's content.
  repeated int32  pub_blocked_cat    = 20;       // Ad categories blocked by the
                                                 // publisher on this request
  repeated int32  pub_blocked_content= 21;       // Ad content blocked by the
                                                 // publisher on this request
  repeated int32  pub_blocked_type   = 22;       // Ad types blocked by the
                                                 // publisher on this request
  repeated string pub_blocked_url    = 23;       // Domains that are blocked
                                                 // from the ad's click URL on
                                                 // this request.
  optional string rtb_data           = 24;       // opaque blob from retargeting
  repeated string pub_allowed_ad_languages = 25; // languages allowed in ads
                                                 // on publishers website
  optional string market_operator    = 26;
  optional Device device             = 27;       // Device Object
  optional bool   is_test            = 28;       // Used for testing.
                                                 // Bidders are expected to return 
                                                 // responses in valid format, which
                                                 // will be used for validation only
                                                 // and get discarded.
}

message BidResponse {
  // ------------------------------------------------------------------------
  // ***    REQUIRED FIELDS                                               ***
  // ------------------------------------------------------------------------
  // These fields are required to be set in the response from the bidder.

  // OPENX identifiers
  required int32  api_version        = 1;  // This value represents the version
                                           // of the API the Market is using
                                           // to communicate with the real-time
                                           // bidder.
                                           // NOTE: this should match the
                                           // version in the BidRequest
  required string auction_id         = 2;  // This is a unique value
                                           // generated by the Market
                                           // and used to identify the
                                           // bid request.
                                           // NOTE: this should match the
                                           // request_id in the BidRequest


  // deprecated, should be in Bid sub-message, see documentation for details
  optional AdId   DEPRECATED_matching_ad_id = 3;
  optional int64  DEPRECATED_cpm_bid_micros = 4;
  optional string DEPRECATED_ad_code  = 5;
  optional string DEPRECATED_click_url = 6;
  optional string DEPRECATED_advertiser_id = 7;
  repeated int32  DEPRECATED_ad_ox_cats = 8;
  repeated string DEPRECATED_click_through_urls = 9;
  optional string DEPRECATED_buyer_id     = 10;
  optional string DEPRECATED_brand_id    = 11;

  // ------------------------------------------------------------------------
  // ***    OPTIONAL FIELDS                                               ***
  // ------------------------------------------------------------------------
  // These fields are optional and may be set by the bidder.
  repeated Bid    bids                   = 12; // list of bids for the given
                                               // auction
  optional int64  next_highest_bid_micros= 13; // the value of the next highest
                                               // bid after those in this
                                               // response
}

message Bid {
  // ------------------------------------------------------------------------
  // ***    REQUIRED FIELDS                                               ***
  // ------------------------------------------------------------------------
  // these fields must be passed with a valid bid, bids of zero or no bids
  // should be left out of the response
  required AdId   matching_ad_id     = 1;  // the id of the matching ad
                                           // selected
  required int64  cpm_bid_micros     = 2;  // The CPM bid expressed in micros.
                                           // Bids must be made in billable
                                           // units (so for USD they must be
                                           // multiples of 10,000).
  required string ad_code            = 3;  // The HTML code to serve if the
                                           // bidder wins the auction.

  // ------------------------------------------------------------------------
  // ***    OPTIONAL FIELDS                                               ***
  // ------------------------------------------------------------------------
  // These fields are optional and may be set by the bidder.

  repeated int32  ad_ox_cats         = 4;  // OX categories of the ad
  repeated string click_through_urls = 5;  // Destination URL of the ad
  optional string buyer_id           = 6;  // the bidder's ID for the primary
                                           // buyer in their system (which is
                                           // often a network or agency)
  optional string brand_id           = 7;  // the bidder's ID for the actual
                                           // underlying brand who's ad will
                                           // show
}

message AuctionResults {
  // ------------------------------------------------------------------------
  // ***    REQUIRED FIELDS                                               ***
  // ------------------------------------------------------------------------
  // These fields are guaranteed to be set by the OpenX RTB system and passed
  // to the bidder in the notify request.

  // OPENX identifiers
  required int32 api_version         = 1;        // This value represents the
                                                 // version of the API the
                                                 // Market is using to
                                                 // communicate with the
                                                 // real-time bidder.
  required string auction_id         = 2;        // This is a unique value
                                                 // generated by the Market
                                                 // and used to identify the
                                                 // auction request.
  repeated AuctionResult results     = 3;        // Information about the
                                                 // result of the auction
}

enum AuctionStatus {
  win = 1;     // the auction was won
  loss = 2;    // the auction was lost
  error = 3;   // an error occured
}

message AuctionResult {
  required AdId          matching_ad_id = 1;  // the ad id for this notification
  required AuctionStatus status = 2;          // win, loss or error
  optional string        loss_reason = 3;     // the reason for the loss, for
                                              // example, "price",
                                              // "disqualification"
                                              // see documentation for more
  optional string        error_reason = 4;    // the reason for the error
                                              // for example, "timeout"
  optional int64         winning_bid_micros = 5; // only set for status loss
  optional int64         clearing_price_micros = 6; // only set for status win
}

